hudi-spark-datasource/hudi-spark/src/test/java/org/apache/hudi/functional/TestGlobalIndexEnableUpdatePartitions.java

```diff
@@ -0,0 +1,224 @@
+import static org.apache.hudi.testutils.Assertions.assertNoWriteErrors;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TestGlobalIndexEnableUpdatePartitions extends SparkClientFunctionalTestHarness {
+
+  private static Stream<Arguments> getTableTypeAndIndexType() {
+    return Stream.of(
+        Arguments.of(COPY_ON_WRITE, GLOBAL_SIMPLE),
+        Arguments.of(COPY_ON_WRITE, GLOBAL_BLOOM),
+        Arguments.of(MERGE_ON_READ, GLOBAL_SIMPLE),
+        Arguments.of(MERGE_ON_READ, GLOBAL_BLOOM)
+    );
+  }
+
+  @ParameterizedTest
+  @MethodSource("getTableTypeAndIndexType")
+  public void testPartitionChanges(HoodieTableType tableType, HoodieIndex.IndexType indexType) throws IOException {
+    final Class<?> payloadClass = DefaultHoodieRecordPayload.class;
+    HoodieWriteConfig writeConfig = getWriteConfig(payloadClass, indexType);
+    HoodieTableMetaClient metaClient = getHoodieMetaClient(tableType, writeConfig.getProps());
+    try (SparkRDDWriteClient client = getHoodieWriteClient(writeConfig)) {
+      final int totalRecords = 4;
+      final String p1 = "p1";
+      final String p2 = "p2";
+      final String p3 = "p3";
+      final String p4 = "p4";
+
+      // 1st batch: inserts
+      String commitTimeAtEpoch0 = getCommitTimeAtUTC(0);
+      List<HoodieRecord> insertsAtEpoch0 = getInserts(totalRecords, p1, 0, payloadClass);
+      client.startCommitWithTime(commitTimeAtEpoch0);
+      assertNoWriteErrors(client.upsert(jsc().parallelize(insertsAtEpoch0, 2), commitTimeAtEpoch0).collect());
+
+      // 2nd batch: normal updates same partition
+      String commitTimeAtEpoch5 = getCommitTimeAtUTC(5);
+      List<HoodieRecord> updatesAtEpoch5 = getUpdates(insertsAtEpoch0, 5, payloadClass);
+      client.startCommitWithTime(commitTimeAtEpoch5);
+      assertNoWriteErrors(client.upsert(jsc().parallelize(updatesAtEpoch5, 2), commitTimeAtEpoch5).collect());
+      readTableAndValidate(metaClient, new int[] {0, 1, 2, 3}, p1, 5);
+
+      // 3rd batch: update all from p1 to p2
+      String commitTimeAtEpoch6 = getCommitTimeAtUTC(6);
+      List<HoodieRecord> updatesAtEpoch6 = getUpdates(updatesAtEpoch5, p2, 6, payloadClass);
+      client.startCommitWithTime(commitTimeAtEpoch6);
+      assertNoWriteErrors(client.upsert(jsc().parallelize(updatesAtEpoch6, 2), commitTimeAtEpoch6).collect());
+      readTableAndValidate(metaClient, new int[] {0, 1, 2, 3}, p2, 6);
+
+      // 4th batch: update all from p2 to p3
+      String commitTimeAtEpoch7 = getCommitTimeAtUTC(7);
+      List<HoodieRecord> updatesAtEpoch7 = getUpdates(updatesAtEpoch6, p3, 7, payloadClass);
+      client.startCommitWithTime(commitTimeAtEpoch7);
+      assertNoWriteErrors(client.upsert(jsc().parallelize(updatesAtEpoch7, 2), commitTimeAtEpoch7).collect());
+      readTableAndValidate(metaClient, new int[] {0, 1, 2, 3}, p3, 7);
+
+      // 5th batch: late update all to p4; discarded
```

Comment:
why discarded? not sure I understand

Reply:
meaning this batch has an older ordering value and won't update the table (discarded).